### PR TITLE
Initial artix7 (using ROI) support for VPR

### PR DESCRIPTION
**Pull request by litghost**
This is missing a couple things, and requires a manual step (design.txt) I haven't been able to figure out how to resolve yet.

Work to be done:
- Supporting multi layouts and ROIs will not possible in the current format.  Need to refactor how prjxray generates ROI's, and think about how to support multiple ROI's.  The code already understands how to generate files for an ROI, but that information is currently in the database, rather than a standalone file.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -7,3 +7,7 @@
 [submodule "third_party/icestorm"]
 	path = third_party/icestorm
 	url = https://github.com/cliffordwolf/icestorm.git
+[submodule "third_party/vtr-verilog-to_routing"]
```
[the remark this answers is not in the document]
Comment (litghost) [reply .gitmodules]: That is intentional.

File: artix7/primitives/slicel/Nlut/ntemplate.Nlut.pb_type.xml [hunk 2]
```diff
@@ -26,13 +26,19 @@
...
       <direct name="{N}LUT_A2_0" input="BLK_IG-{N}LUT.A2"      output="BEL_LT-{N}5LUT[0].in[1]"/><direct name="{N}LUT_A2_1" input="BLK_IG-{N}LUT.A2"      output="BEL_LT-{N}5LUT[1].in[1]"/>
       <direct name="{N}LUT_A1_0" input="BLK_IG-{N}LUT.A1"      output="BEL_LT-{N}5LUT[0].in[0]"/><direct name="{N}LUT_A1_1" input="BLK_IG-{N}LUT.A1"      output="BEL_LT-{N}5LUT[1].in[0]"/>
       <direct name="O5"          input="BEL_LT-{N}5LUT[0].out" output="BLK_IG-{N}LUT.O5"       /><direct name="O6"          input="BEL_LT-{N}5LUT[1].out" output="BLK_IG-{N}LUT.O6"       />
     </interconnect>
+    <metadata>
+      <meta name="fasm_lut">
+        {N}LUT.INIT
+        {N}LUT.SMALL
+      </meta>
+    </metadata>
```
[the remark this answers is not in the document]
Comment (litghost) [reply artix7/primitives/slicel/Nlut/ntemplate.Nlut.pb_type.xml]: Ya, I like this

File: artix7/primitives/slicel/slicel.pb_type.xml [hunk 3]
```diff
@@ -141,30 +169,116 @@
...
-      output="BLK_IG-SLICEL.DMUX" />
+      output="BLK_IG-SLICEL.DMUX">
+      <metadata>
+        <meta name="fasm_mux">
+          BLK_IG-D5FF.Q => DMUX.D5Q
+          BLK_IG-DLUT.O5 => DMUX.O5
+          BLK_IG-DLUT.O6 => DMUX.O6
+          BLK_IG-CARRY[3].CO => DMUX.CY
+          BLK_IG-CARRY[3].O => DMUX.XOR
+        </meta>
```
[the remark this answers is not in the document]
Comment (litghost) [reply artix7/primitives/slicel/slicel.pb_type.xml]: No, if BLK_IG-DFF.Q path enable, then connect input of mux DMUX.D5Q to output of mux

File: artix7/primitives/slicel/carry/carry.pb_type.xml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+  <interconnect>
+    <direct input="BLK_IG-CARRY.CIN" name="CIN_TO_MUXCY" output="BEL_MX-MUXCY.CI">
+      <pack_pattern in_port="BLK_IG-CARRY.CIN" name="CARRYCHAIN" out_port="BEL_MX-MUXCY.CI"/>
+    </direct>
+    <direct input="BLK_IG-CARRY.DI" name="DI_TO_MUXCY" output="BEL_MX-MUXCY.DI"/>
+    <direct input="BLK_IG-CARRY.S" name="S_TO_MUXCY" output="BEL_MX-MUXCY.S"/>
+
+    <direct input="BEL_MX-MUXCY.CO" name="MUXCY_TO_COUT" output="BLK_IG-CARRY.COUT">
+      <pack_pattern in_port="BEL_MX-MUXCY.CO" name="CARRYCHAIN" out_port="BLK_IG-CARRY.COUT"/>
```
[the remark this answers is not in the document]
Comment (litghost) [reply artix7/primitives/slicel/carry/carry.pb_type.xml]: Seems excessively verbose

File: artix7/utils/prjxray-arch-import.py [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+        else:
+            # We don't want this tile
+            continue
+
+        single_xml = ET.SubElement(fixed_layout_xml, 'single', {
+                'priority': '1',
+                'type': vpr_tile_type,
+                'x': str(loc[0]),
+                'y': str(loc[1]),
+        })
```
[the remark this answers is not in the document]
Comment (litghost) [reply artix7/utils/prjxray-arch-import.py]: Agreed

File: artix7/primitives/slicel/slicel.pb_type.xml [hunk 6]
```diff
@@ -141,30 +169,116 @@
...
       input="BLK_IG-CARRY[3].O BLK_IG-CARRY[3].CO BLK_IG-DLUT.O6 BLK_IG-DLUT.O5 BLK_IG-SLICEL.DX"
-      output="BLK_BB-DFF.D" />
+      output="BLK_BB-DFF.D" >
+      <metadata>
+        <meta name="fasm_mux">
+          BLK_IG-SLICEL.DX => DFF.DMUX.DX
+          BLK_IG-DLUT.O5 => DFF.DMUX.O5
+          BLK_IG-DLUT.O6 => DFF.DMUX.O6
+          BLK_IG-CARRY[3].CO => DFF.DMUX.CY
+          BLK_IG-CARRY[3].O => DFF.DMUX.XOR
```
[the remark this answers is not in the document]
Comment (litghost) [reply artix7/primitives/slicel/slicel.pb_type.xml]: Nah, explicit over implicit.  Take for example:
  BLK_IG-DLUT.O5 => D5FF.MUX.A                                                               
  BLK_IG-SLICEL.DX => D5FF.MUX.B

File: artix7/primitives/slicel/Nlut/ntemplate.Nlut.pb_type.xml [hunk 7]
```diff
@@ -56,5 +62,10 @@
       <direct name="{N}LUT_A1" input="BLK_IG-{N}LUT.A1"   output="BEL_LT-{N}6LUT.in[0]"/>
       <direct name="O6"        input="BEL_LT-{N}6LUT.out" output="BLK_IG-{N}LUT.O6"/>
     </interconnect>
+    <metadata>
+      <meta name="fasm_lut">
+        {N}LUT.INIT
+      </meta>
+    </metadata>
```
[the remark this answers is not in the document]
Comment (litghost) [reply artix7/primitives/slicel/Nlut/ntemplate.Nlut.pb_type.xml]: the fasm_enable needs to be on the mode, and the fasm_type actually varies by mode.

File: artix7/utils/prjxray-arch-import.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mithro) [reply artix7/utils/prjxray-arch-import.py]: I'm unsure if I got as far as allowing metadata on the grid definitions. I think we may have just ended up hard coding the tile name generation...

[the remark this answers is not in the document]
Comment (litghost) [reply artix7/utils/prjxray-arch-import.py]: Worked, I just added the rest of the FASM metadata.

File: artix7/primitives/slicel/slicel.pb_type.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (mithro) [reply artix7/primitives/slicel/slicel.pb_type.xml]: I would like to avoid encoding "another" format inside XML. We already have XML and the weird port/bit syntax plus array syntax.

The metadata stuff seems to already support the concept of a value at an offset on the rr_nodes. Could we reuse that?

If I recall correctly,
```xml
 <mux name="bla" input="multibit.port" output="blah">
```
Is the same as 
```xml
 <mux name="bla" input="multibit.port[0] multibit.port[1] ... Multibit.port[n]" output="blah">
```

**Comment by mithro** [comment]
@litghost 
Can you do some minor clean up to this pull request;
 - [ ] Rebase onto master (removing merge commits).
 - [ ] Remove my bram / slicem changes.
 - [ ] Remove the flat buffer stuff from the history.
 - [ ] (maybe) Squash the revert or reverts together?

File: artix7/CMakeLists.txt [hunk 8]
```diff
@@ -1,22 +1,53 @@
...
+
+
 add_subdirectory(tiles)
 add_subdirectory(blocks)
 
+
 define_arch(
   ARCH artix7
-  YOSYS_SCRIPT "synth_xilinx -vpr"
+  YOSYS_SCRIPT "synth_xilinx -vpr -flatten"
```
Comment (mithro) [inline artix7/CMakeLists.txt]: Can you add a comment about why `-flatten` is needed?

File: artix7/make/project_xray.cmake [hunk 9]
```diff
@@ -1,86 +1,248 @@
...
 
-  set(CLB_IMPORT ${symbiflow-arch-defs_SOURCE_DIR}/artix7/utils/prjxray-clb-import.py)
-  get_project_xray_dependencies(DEPS ${PROJECT_XRAY_CLB_PART} ${CLB})
+  set(TILE_IMPORT ${symbiflow-arch-defs_SOURCE_DIR}/artix7/utils/prjxray-tile-import.py)
+  get_project_xray_dependencies(DEPS ${PROJECT_XRAY_TILE_PART} ${TILE})
+
+  set(PART ${PROJECT_XRAY_TILE_PART})
+  set(PB_TYPE_INCLUDE_FILES "")
+  set(MODEL_INCLUDE_FILES "")
+  foreach(SITE_TYPE ${PROJECT_XRAY_TILE_SITE_TYPES})
```
Comment (mithro) [inline artix7/make/project_xray.cmake]: I think site types should be passed to the `tile_import` tool. That way it can check that the site type is valid for this tile type **and** skip site types which are not being provided.

File: artix7/primitives/ff/ff.pb_type.xml [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+          <direct name="CLR" input="BEL_FF-FDPE_or_FDCE.SR" output="BEL_FF-FDCE.CLR" />
+          <direct name="Q" input="BEL_FF-FDCE.Q" output="BEL_FF-FDPE_or_FDCE.Q" />
+        </interconnect>
+
+        <metadata>
+          <meta name="fasm_features">ZRST</meta>
+        </metadata>
+      </mode>
+      <metadata>
+        <meta name="fasm_prefix">AFF BFF CFF DFF A5FF B5FF C5FF D5FF</meta>
```
Comment (mithro) [inline artix7/primitives/ff/ff.pb_type.xml]: The idea here is that each pb_type in the array is a given value in the `fasm_prefix`?

File: artix7/primitives/ff/ff.pb_type.xml [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+
+      <direct name="D" input="BLK_BB-SLICE_FF.D[3:0]" output="BEL_FF-FDPE_or_FDCE[3:0].D" />
+      <direct name="Q" input="BEL_FF-FDPE_or_FDCE[3:0].Q" output="BLK_BB-SLICE_FF.Q[3:0]" />
+
+      <direct name="D5" input="BLK_BB-SLICE_FF.D5[3:0]" output="BEL_FF-FDPE_or_FDCE[7:4].D" />
+      <direct name="Q5" input="BEL_FF-FDPE_or_FDCE[7:4].Q" output="BLK_BB-SLICE_FF.Q5[3:0]" />
+    </interconnect>
+  </mode>
+  <mode name="LDPE_or_LDCE">
+    <pb_type name="BEL_FF-LDPE_or_LDCE" num_pb="8">
```
Comment (mithro) [inline artix7/primitives/ff/ff.pb_type.xml]: I was pretty sure that only *some* of the FFs supported working in latch mode?

Comment (mithro) [reply artix7/primitives/ff/ff.pb_type.xml]: image

File: artix7/primitives/slicel/N5ff/ntemplate.N5ff.sim.v [hunk 12]
```diff
@@ -1,53 +0,0 @@
-`include "../../../../vpr/ff/vpr_ff.sim.v"
```
Comment (mithro) [inline artix7/primitives/slicel/N5ff/ntemplate.N5ff.sim.v]: Can we keep this file around somewhere?

File: artix7/primitives/ff/ff.pb_type.xml [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+
+      <direct name="D5" input="BLK_BB-SLICE_FF.D5[3:0]" output="BEL_FF-FDSE_or_FDRE[7:4].D" />
+      <direct name="Q5" input="BEL_FF-FDSE_or_FDRE[7:4].Q" output="BLK_BB-SLICE_FF.Q5[3:0]" />
+    </interconnect>
+    <metadata>
+      <meta name="fasm_features">FFSYNC</meta>
+    </metadata>
+  </mode>
+  <mode name="FDPE_or_FDCE">
+    <pb_type name="BEL_FF-FDPE_or_FDCE" num_pb="8">
```
Comment (mithro) [inline artix7/primitives/ff/ff.pb_type.xml]: I think you actually want two sets of `<pb_type>` objects here, 4 * `nFF` and 4 * `n5FF`.

File: artix7/primitives/slicel/carry4_modes/carry4/carry4.sim.v [hunk 14]
```diff
@@ -1,26 +0,0 @@
-(* blackbox *)
-(* MODEL_NAME="CARRY4" *)
```
Comment (mithro) [inline artix7/primitives/slicel/carry4_modes/carry4/carry4.sim.v]: Can we keep these `.sim.v` files somewhere for now?

File: artix7/primitives/slicel/slicel.pb_type.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (mithro) [reply artix7/primitives/slicel/slicel.pb_type.xml]: I still dislike the current syntax, it's encoding another format inside an XML document.

File: artix7/tests/buttons/buttons.pcf [hunk 15]
```diff
@@ -0,0 +1,17 @@
+set_io in[0] V17
```
Comment (mithro) [inline artix7/tests/buttons/buttons.pcf]: Can you add comments to this file?

I assume the top set is the buttons on the basys and the bottom set are the LEDs on the basys?

File: artix7/tests/murax/Murax.v [hunk 16]
```diff
@@ -0,0 +1,4558 @@
+// Generator : SpinalHDL v1.1.5    git head : 0310b2489a097f2b9de5535e02192d9ddd2764ae
```
Comment (mithro) [inline artix7/tests/murax/Murax.v]: Can we share the murax test between the ice40 and the artix7?

File: artix7/tests/picosoc/README.md [hunk 17]
```diff
@@ -0,0 +1,8 @@
+# PicoSoC
+
```
Comment (mithro) [inline artix7/tests/picosoc/README.md]: Can we share PicoSoC between ice40 and artix7?

File: artix7/utils/prjxray-arch-import.py [hunk 18]
```diff
@@ -0,0 +1,334 @@
+import argparse
```
Comment (mithro) [inline artix7/utils/prjxray-arch-import.py]: Docstring at top of file describing what it does.

File: artix7/utils/prjxray-arch-import.py [hunk 19]
```diff
@@ -0,0 +1,334 @@
...
+import prjxray.db
+import os.path
+import simplejson as json
+import csv
+
+import lxml.etree as ET
+
+def create_synth_io_tiles(complexblocklist_xml, pb_name, is_input):
```
Comment (mithro) [inline artix7/utils/prjxray-arch-import.py]: Add a docstring comment to this function.

File: artix7/utils/prjxray-tile-import.py [hunk 20]
```diff
@@ -0,0 +1,301 @@
+#!/usr/bin/env python3
+
+"""
+Import the top level tile interconnect information from Project X-Ray database
+files.
```
Comment (mithro) [inline artix7/utils/prjxray-tile-import.py]: This doc could do with a little more expansion....

File: artix7/utils/prjxray_assign_tile_pin_direction.py [hunk 21]
```diff
@@ -0,0 +1,289 @@
+import argparse
```
Comment (mithro) [inline artix7/utils/prjxray_assign_tile_pin_direction.py]: File comment.

File: artix7/utils/prjxray_create_ioplace.py [hunk 22]
```diff
@@ -0,0 +1,88 @@
+from __future__ import print_function
```
Comment (mithro) [inline artix7/utils/prjxray_create_ioplace.py]: File comment.

File: artix7/utils/prjxray_form_channels.py [hunk 23]
```diff
@@ -0,0 +1,243 @@
+#!/usr/bin/env python3
```
Comment (mithro) [inline artix7/utils/prjxray_form_channels.py]: File comment.

File: artix7/utils/prjxray-tile-import.py [hunk 24]
```diff
@@ -0,0 +1,301 @@
+#!/usr/bin/env python3
```
Comment (mithro) [inline artix7/utils/prjxray-tile-import.py]: Why is this `prjxray-tile-import.py` but everything else seems to be `prxray_my_face.py`?

File: artix7/utils/prjxray_routing_import.py [hunk 25]
```diff
@@ -0,0 +1,416 @@
+#!/usr/bin/env python3
+
```
Comment (mithro) [inline artix7/utils/prjxray_routing_import.py]: File comment.

File: artix7/utils/prjxray_routing_compare.py [hunk 26]
```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env python3
+
```
Comment (mithro) [inline artix7/utils/prjxray_routing_compare.py]: File comment.

File: artix7/utils/prjxray_synth_tiles_to_pinmap_csv.py [hunk 27]
```diff
@@ -0,0 +1,52 @@
+from __future__ import print_function
+import argparse
```
Comment (mithro) [inline artix7/utils/prjxray_synth_tiles_to_pinmap_csv.py]: File comment.

File: artix7/utils/prjxray_synth_tiles_to_pinmap_csv.py [hunk 28]
```diff
@@ -0,0 +1,52 @@
+from __future__ import print_function
```
Comment (mithro) [inline artix7/utils/prjxray_synth_tiles_to_pinmap_csv.py]:
```suggestion
#! /usr/bin/env python3
from __future__ import print_function
```

File: ice40/utils/ice40_create_ioplace.py [hunk 29]
```diff
@@ -1,12 +1,13 @@
 #!/usr/bin/env python3
```
Comment (mithro) [inline ice40/utils/ice40_create_ioplace.py]: File comment (maybe?).

File: make/devices.cmake [hunk 30]
```diff
@@ -363,10 +403,12 @@ function(DEFINE_DEVICE)
         ${QUIET_CMD} ${VPR} ${DEVICE_MERGED_FILE}
         --device ${DEVICE_FULL}
         ${symbiflow-arch-defs_SOURCE_DIR}/common/wire.eblif
-        --route_chan_width 100
+        --route_chan_width 6
```
Comment (mithro) [inline make/devices.cmake]: Why 6?

File: make/devices.cmake [hunk 31]
```diff
@@ -363,10 +403,12 @@ function(DEFINE_DEVICE)
...
         --echo_file on
         --min_route_chan_width_hint 1
```
Comment (mithro) [inline make/devices.cmake]: I seem to recall things are faster if this matched `--route_chan_width`?

File: testarch/CMakeLists.txt [hunk 32]
```diff
@@ -14,6 +14,7 @@ define_arch(
   --write_rr_graph \${OUT_RRXML_REAL}"
   NO_PINS
   NO_BITSTREAM
+  ROUTE_CHAN_WIDTH 100
```
Comment (mithro) [inline testarch/CMakeLists.txt]: Why is this needed? Everything needs a `ROUTE_CHAN_WIDTH` now?

File: utils/lib/rr_graph/channel2.py [hunk 33]
```diff
@@ -0,0 +1,33 @@
+from intervaltree import IntervalTree, Interval
```
Comment (mithro) [inline utils/lib/rr_graph/channel2.py]: File comment.

Why do we have both `channel.py` and `channel2.py`.

File: utils/lib/rr_graph/graph2.py [hunk 34]
```diff
@@ -0,0 +1,437 @@
+from collections import namedtuple
```
Comment (mithro) [inline utils/lib/rr_graph/graph2.py]: File comment.

Why is `graph2.py` existing?

File: utils/lib/rr_graph/points.py [hunk 35]
```diff
@@ -12,8 +12,8 @@
 from collections import defaultdict
 from collections import namedtuple
 
-from ..asserts import assert_eq
-from ..asserts import assert_type
+from lib.asserts import assert_eq
```
Comment (mithro) [inline utils/lib/rr_graph/points.py]: These should remain as relative imports -- the file has other relative imports?

File: utils/lib/rr_graph_fb/flatbuffers.py [hunk 36]
```diff
@@ -0,0 +1,26 @@
+""" These functions should be moved to the flatbuffers.py library itself."""
```
Comment (mithro) [inline utils/lib/rr_graph_fb/flatbuffers.py]: Remove the `rr_graph_fb` from this patch. Lets land that in a different one.

File: utils/mux_gen.py [hunk 37]
```diff
@@ -72,7 +72,7 @@
 )
 
 parser.add_argument(
-    '--name-output',
+    '--name-outputs',
```
Comment (mithro) [inline utils/mux_gen.py]: Why do you need `--name-outputs` -- IE there should only be one output on a mux right?

File: artix7/primitives/ff/ff.pb_type.xml (hunk 11 above)
Comment (litghost) [reply artix7/primitives/ff/ff.pb_type.xml]: Yep, that is correct.  Sad :(

File: artix7/make/project_xray.cmake (hunk 9 above)
Comment (litghost) [reply artix7/make/project_xray.cmake]: Agreed/

File: artix7/CMakeLists.txt (hunk 8 above)
Comment (litghost) [reply artix7/CMakeLists.txt]: Because if it isn't flattened, the eblif has multiple modules, which confuses things.  ice40 currently uses -flatten too.

File: artix7/primitives/slicel/slicel.pb_type.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (litghost) [reply artix7/primitives/slicel/slicel.pb_type.xml]: That's actually true of all the FASM metadata, so I'm not using that as a bar to cross.  I don't think the LUT metadata should be XML e.g. `<lut start_bit="4" end_bit="6">LUT.INIT</lut>` versus LUT.INIT[6:4].

File: artix7/primitives/slicel/N5ff/ntemplate.N5ff.sim.v (hunk 12 above)
Comment (litghost) [reply artix7/primitives/slicel/N5ff/ntemplate.N5ff.sim.v]: It's in the git history?  That's somewhere.

File: utils/mux_gen.py (hunk 37 above)
Comment (litghost) [reply utils/mux_gen.py]: For a minute I wanted a 2 output MUXCY, I'll revert this.

File: utils/lib/rr_graph_fb/flatbuffers.py (hunk 36 above)
Comment (litghost) [reply utils/lib/rr_graph_fb/flatbuffers.py]: Done

File: testarch/CMakeLists.txt (hunk 32 above)
Comment (litghost) [reply testarch/CMakeLists.txt]: Because the required ROUTE_CHAN_WIDTH varies by arch.

File: make/devices.cmake (hunk 30 above)
Comment (litghost) [reply make/devices.cmake]: It's even and small.  Emitting, and filtering a virtual RR graph with channel width 100 is much much slow than a 6 width channel.  The virtual RR graph is pretty much only used for seeding the real RR graph.

File: artix7/primitives/slicel/slicel.pb_type.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (mithro) [reply artix7/primitives/slicel/slicel.pb_type.xml]: FYI - `LUT.INIT[6:4]` is an existing format that can be used in a number of other places in the XML file.

File: artix7/tests/picosoc/README.md (hunk 17 above)
Comment (litghost) [reply artix7/tests/picosoc/README.md]: Sure, but for now I'll just remove the projects and add back in another pull request.

**Comment by mithro** [comment]
```
[ 16%] Generating buttons/artix7-xc7a50t-roi-virt-xc7a50t-test/top.net

VPR FPGA Placement and Routing.
Version: 8.0.0-dev+vpr-7.0.5-6098-g1169f2e67
Revision: vpr-7.0.5-6098-g1169f2e67
Compiled: 2018-10-25T05:26:08 (release build)
Compiler: GNU 7.3.0 on Linux-4.14.12-041412-generic x86_64
University of Toronto
vtr-users@googlegroups.com
This is free open source code under MIT license.

Argument --route_chan_width specified multiple times

usage: vpr architecture circuit [--pack] [--place] [--route] [--analysis]
       [--disp {on, off}] [-h] [--version] [--device DEVICE_NAME]
       [-j NUM_WORKERS] [--timing_analysis {on, off}]
       [--route_chan_width CHANNEL_WIDTH] [OTHER_OPTIONS ...]
artix7/tests/buttons/CMakeFiles/buttons_bit.dir/build.make:123: recipe for target 'artix7/tests/buttons/buttons/artix7-xc7a50t-roi-virt-xc7a50t-test/top.net' failed
make[3]: *** [artix7/tests/buttons/buttons/artix7-xc7a50t-roi-virt-xc7a50t-test/top.net] Error 1
make[3]: Leaving directory '/big/github/SymbiFlow/symbiflow-arch-defs/build'
CMakeFiles/Makefile2:54219: recipe for target 'artix7/tests/buttons/CMakeFiles/buttons_bit.dir/all' failed
make[2]: *** [artix7/tests/buttons/CMakeFiles/buttons_bit.dir/all] Error 2
make[2]: Leaving directory '/big/github/SymbiFlow/symbiflow-arch-defs/build'
Makefile:138: recipe for target 'all' failed
make[1]: *** [all] Error 2
make[1]: Leaving directory '/big/github/SymbiFlow/symbiflow-arch-defs/build'
Makefile:4: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by litghost** [comment]
> Argument --route_chan_width specified multiple times

This is due to a stale CMakeCache.  I moved the argument from one cached CMake variable to another.  You can either edit you CMakeCache (not recommended) or just delete the file (its in the root of your build directory) and run cmake <source dir> again.

**Comment by mithro** [comment]
Why isn't cmake removing the CMakeCache?

**Comment by litghost** [comment]
> Why isn't cmake removing the CMakeCache?

Because that is the behavior of the CMakeCache.  The is specifically user controlled.  Read the "Variable types in CMake" section in https://cmake.org/cmake/help/v3.0/command/set.html
